### PR TITLE
Build and publish C++ and Slice API ref

### PR DESCRIPTION
**Pull request by externl**
This PR adds a new workflow to build (main and PRs) and publish (main only) the Slice and C++ API reference. 

Due to limitations of the S3 api (no hash based syncing, only time + size checks) we do a full sync every commit. I think this is fine since, we're not committing that much, and as per the S3 the pricing is reasonable.

https://aws.amazon.com/s3/pricing/

```
Data Transfer IN To Amazon S3 From Internet	
All data transfer in	$0.00 per GB
```

This PR also fixes a number of bugs that were causing the Doxygen builds to fail.

### CODE REVIEW

File: .github/workflows/docs.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
+name: Generate documentation
+
+on:
+  workflow_dispatch:
+  push:
+    branches: ["main", "publish-api-reference"]
```
Comment (pepone) [inline .github/workflows/docs.yml]: What is the "publish-api-reference" branch for?

File: .github/workflows/docs.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  pull_request:
+    # The branches below must be a subset of the branches above
+    branches: ["main"]
+
+jobs:
+  ci:
+    name: Build Doxygen documentation
+
+    runs-on: macos-14
```
Comment (pepone) [inline .github/workflows/docs.yml:15]: Why don't use Ubuntu? aren't macOS VMs concurrent jobs limited?

File: .github/workflows/docs.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+          make generate-srcs
+          cd doxygen
+          doxygen
+
+      # This will perform a full sync of the documentation to S3 every time the workflow is run since
+      # the timestamps will always be different. Using --size-only is not sufficient since the
+      # documentation may be updated without changing the size of the files. S3 does not offer a hash based sync.
+      #
+      # Additionally, we do not cache the doxygen output since it does not remove files old files.
+      - name: Sync Documentation to S3
```
Comment (pepone) [inline .github/workflows/docs.yml:47]: why are we publishing the docs for main on S3?

File: .github/workflows/docs.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+          aws s3 cp ./doxygen/slice.tag s3://${AWS_S3_DOC_BUCKET}/api/ice/main/slice.tag
+
+          aws s3 sync ./cpp/doxygen/cpp s3://${AWS_S3_DOC_BUCKET}/api/ice/main/cpp --delete
+          aws s3 cp ./cpp/doxygen/icecpp.tag s3://${AWS_S3_DOC_BUCKET}/api/ice/main/icecpp.tag
+        env:
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          AWS_S3_DOC_BUCKET: ${{ secrets.AWS_S3_DOC_BUCKET }}
+          AWS_DEFAULT_REGION: us-east-1
+        # if: github.ref == 'refs/heads/main'
```
Comment (pepone) [inline .github/workflows/docs.yml]: Should we uncomment this?

**Review by pepone — APPROVED** [review]

File: .github/workflows/docs.yml (hunk 1 above)
Comment (externl) [reply .github/workflows/docs.yml]: that's the one I was testing on. Will remove.

File: .github/workflows/docs.yml (hunk 3 above)
Comment (externl) [reply .github/workflows/docs.yml:47]: The plan is it to have have the docs public for the for various branches. No need for our internal site.

We you can always see the latest api ref at:

https://doc.zeroc.com/api/ice/main/cpp/index.html

File: .github/workflows/docs.yml (hunk 4 above)
Comment (externl) [reply .github/workflows/docs.yml]: Yes, I've pushed a fix.

**Pull request merged**